src/value.jl

```diff
@@ -1,145 +1,174 @@
-    return CategoricalValue(convert(R, level), pool)
-end
-
-Base.convert(::Type{CategoricalValue{T, R}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{CategoricalValue{T}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{CategoricalValue}, x::CategoricalValue) = x
-
-Base.convert(::Type{Union{CategoricalValue{T, R}, Null}}, x::CategoricalValue{T, R}) where {T, R <: Integer} = x
-Base.convert(::Type{Union{CategoricalValue{T}, Null}}, x::CategoricalValue{T}) where {T} = x
-Base.convert(::Type{Union{CategoricalValue, Null}}, x::CategoricalValue) = x
-
-Base.promote_rule(::Type{CategoricalValue{S, R}}, ::Type{T}) where {S, T, R} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue{S}}, ::Type{T}) where {S, T} = promote_type(S, T)
-Base.promote_rule(::Type{CategoricalValue}, ::Type{T}) where {T} = T
+# union of all types that have "categorical value" trait
+const CatValue{R} = Union{CategoricalValue{T, R} where T,
+                          CategoricalString{R}}
+
+# checks whether the type is "categorical value"
+iscatvalue(::Type) = false
+iscatvalue(::Type{Union{}}) = false # otherwise it dispatches to Type{<:...}
+iscatvalue(::Type{<:CatValue}) = true
+iscatvalue(x::Any) = iscatvalue(typeof(x))
+
+leveltype(::Type{<:CategoricalValue{T}}) where {T} = T
+leveltype(::Type{<:CategoricalString}) = String
+leveltype(::Type) = throw(ArgumentError("Not a \"categorical value\" type"))
+leveltype(x::Any) = leveltype(typeof(x))
+# eltype() is a synonym of leveltype() for categorical values
+Base.eltype(::Type{T}) where {T <: CatValue} = leveltype(T)
+Base.eltype(x::CatValue) = eltype(typeof(x))
+
+# integer type of category reference codes used by categorical value
+reftype(::Type{<:CatValue{R}}) where {R} = R
+reftype(x::Any) = reftype(typeof(x))
+
+pool(x::CatValue) = x.pool
+level(x::CatValue) = x.level
+
+# extract the type of the original value from array eltype `T`
+unwrap_catvaluetype(::Type{T}) where {T} = T
+unwrap_catvaluetype(::Type{T}) where {T >: Null} =
+    Union{unwrap_catvaluetype(Nulls.T(T)), Null}
+unwrap_catvaluetype(::Type{Null}) = Null # to prevent dispatching to T>:Null method
+unwrap_catvaluetype(::Type{Any}) = Any # to prevent dispatching to T>:Null method
+unwrap_catvaluetype(::Type{T}) where {T <: CatValue} = leveltype(T)
+
+# get the "categorical value" type given value type `T` and reference type `R`
+catvaluetype(::Type{T}, ::Type{R}) where {T >: Null, R} =
+    catvaluetype(Nulls.T(T), R)
+catvaluetype(::Type{T}, ::Type{R}) where {T <: CatValue, R} =
+    reftype(T) === R ? T : catvaluetype(leveltype(T), R)
+catvaluetype(::Type{Any}, ::Type{R}) where {R} =
+    CategoricalValue{Any, R}  # to prevent dispatching to T>:Null method
+catvaluetype(::Type{T}, ::Type{R}) where {T, R} =
+    CategoricalValue{T, R}
+catvaluetype(::Type{<:AbstractString}, ::Type{R}) where {R} =
+    CategoricalString{R}
+# default to CategoricalString in degenerated case (all nulls)
```

Comment:
Repeating my comment from above: This sounds a bit weird. Doesn't it work if you return `CategoricalValue{Union{}, R}`? That's really the correct type, since we know it won't contain anything.

Reply:
I made it consistent with the current behaviour that picks `CategoricalString` for empty arrays with unspecified type:
```julia
CategoricalArray(dims::Int...; ordered=false) =
    CategoricalArray{String}(dims, ordered=ordered)
```

If for all-nulls case we have `CategoricalValue{Union{}}` then for consistency we should have it for empty array as well.

> since we know it won't contain anything.

The type -- yes, but the array could be modified.

Reply:
The default constructor for empty arrays is just a convenience to avoid writing `{String}` explicitly since that's the most common use case. It's really no more than that.

OTC, if the user created an array from an `AbstractArray{Null}`, the array is not supposed to be able to contain anything other than `null`. That's important for example if a column of nulls is concatenated later with a column of non-null values, which could be of any type `T`: the column of nulls should not interfere with type promotion, i.e. `T` should be preserved. That's how it works for e.g. `[null]` already. (That's not just theoretical: I've precisely hit this problem in R a few minutes ago with CSV files containing all-null columns which are considered as being boolean.)

Reply:
> The default constructor for empty arrays is just a convenience to avoid writing `{String}`

But it also leads to an "interesting" behaviour of e.g. `vcat()` (master, but the same holds for this PR):
```
julia> x = CategoricalVector(0)
0-element CategoricalArrays.CategoricalArray{String,1,UInt32}

julia> y = CategoricalVector{Int}(0)
0-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}

julia> z = CategoricalVector([1])
1-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}:
 1

julia> vcat(x, y)
0-element CategoricalArrays.CategoricalArray{Union{Nulls.Null, String},1,UInt32}

julia> vcat(y)
0-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}

julia> vcat(x, z)
1-element CategoricalArrays.CategoricalArray{Union{Int64, Nulls.Null},1,UInt32}:
 1
```

So the result of `vcat()` depends on the emptyness/nonemptyness of its arguments, just like R depends on all-nulls/not all-nulls.
I guess `vcat()` was tuned at some point to workaround the default type of an empty array.

Reply:
Good catch. That's a bug with `CategoricalArrays.mergelevels(false, ["a"], Int[])`, which has a fast path for when a set of levels is empty. We should always respect the original element type even if the array is empty.

Reply:
See https://github.com/JuliaData/CategoricalArrays.jl/pull/89.